# [\#11692 Issue](https://github.com/ethereum/solidity/issues/11692) `closed`: Balance of ERC20 token setting to 0 after try/catch 
**Labels**: `waiting for more input`, `closed due inactivity`, `stale`


#### <img src="https://avatars.githubusercontent.com/u/61954243?u=092f34588505a68339fa458522081fc3c6a05dcb&v=4" width="50">[MartinSchere](https://github.com/MartinSchere) opened issue at [2021-07-25 14:18](https://github.com/ethereum/solidity/issues/11692):

## Description

I'm working on a payment splitter contract that receives LP tokens as well as normal tokens:
- If the token is LP, redeem them using `removeLiquidityETH` and then convert to ETH
- Otherwise, just convert them to ETH

However, I found a very weird bug with the try/catch statement: It's setting the token balance to 0 somehow:

```solidity
function receiveAndDistribute(uint256 amount, address token) external override {
        require(amount > 0, "Distribution: Amount must be greater than 0");
        
        address WETH = router.WETH();
        IERC20(token).transferFrom(address(msg.sender), address(this), amount);
        IERC20(token).approve(address(router), amount);

        uint balanceBefore = IERC20(token).balanceOf(address(this)); // Truffle debugger says it's 10

        // Check if token is LP
        try IUniswapV2Pair(token).getReserves() {
            // Remove liquidity code
        } catch {
            uint balanceAfter = IERC20(token).balanceOf(address(this)); // Now it's 0 according to truffle?!
        }
    }
```

Even if I implement the most simple function:

```solidity
function checkLP(address token) internal view returns(bool){
		bool isLp;
		try IUniswapV2Pair(token).token0(){
			isLp = true;
		} catch {
			isLp = false;
		}
		return isLp;
	}
```
And then use it in the same function:

 ```solidity
 function receiveAndDistribute(uint256 amount, address token) external override {
		require(amount > 0, "Distribution: Amount must be greater than 0");
		
		address WETH = router.WETH();
		IERC20(token).transferFrom(address(msg.sender), address(this), amount);
		IERC20(token).approve(address(router), amount);

		uint balanceBefore = IERC20(token).balanceOf(address(this)); // 10 * 10**18
		address addressBefore = address(this); // 0x28...27577

		bool isLp = checkLP(token); // This is where everything breaks

		address addressAfter = address(this); // 0x28...27577
		uint balanceAfter = IERC20(token).balanceOf(address(this)); // 0
}
```

When I remove the try/catch, the balance is correct and I'm able to call `transferFrom` in the token, otherwise I get `TransferHelper::transferFrom: transferFrom failed` (because the balance is 0)


## Environment

- Compiler version: 0.8.2+commit.661d1103.Emscripten.clang
- Target EVM version (as per compiler settings): istanbul 
- Framework/IDE (e.g. Truffle or Remix): truffle
- EVM execution environment / backend / blockchain client: ganache
- Operating system: WSL2 on Windows 10 pro

## Steps to Reproduce

```solidity
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.2;

import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
import "@uniswap/v2-core/contracts/interfaces/IUniswapV2Pair.sol";
import "@uniswap/v2-periphery/contracts/interfaces/IUniswapV2Router02.sol";
import "@uniswap/v2-periphery/contracts/interfaces/IWETH.sol";


contract Distribution{

	IUniswapV2Router02 router;

	constructor(IUniswapV2Router02 _router){
		router = _router;
	}

	function testBug(uint256 amount, address token) external  {
		IERC20(token).transferFrom(address(msg.sender), address(this), amount);
		IERC20(token).approve(address(router), amount);

		uint balanceBefore = IERC20(token).balanceOf(address(this));

		try IUniswapV2Pair(token).token0(){} catch {}

		uint balanceAfter = IERC20(token).balanceOf(address(this));

	}
}
```

Steps:
- Compile using 0.8.2
- Instantiate the contract with UniswapRouterV2's address
- Try using the function with any ERC20 token
- With a debugger, check the values of `balanceBefore` and `balanceAfter`



#### <img src="https://avatars.githubusercontent.com/u/13174375?u=52d702cb6bec53b561afa293cf9cd53ef7a63924&v=4" width="50">[hrkrshnn](https://github.com/hrkrshnn) commented at [2021-07-26 09:58](https://github.com/ethereum/solidity/issues/11692#issuecomment-886555439):

```solidity
	function testBug(uint256 amount, address token) external  {
		IERC20(token).transferFrom(address(msg.sender), address(this), amount);
		IERC20(token).approve(address(router), amount);

		uint balanceBefore = IERC20(token).balanceOf(address(this));

		try IUniswapV2Pair(token).token0(){} catch {}

		uint balanceAfter = IERC20(token).balanceOf(address(this));

	}
```

Are you saying that `balanceBefore` and `balanceAfter` is different in this case?

It might be easier to discuss this on solidity gitter / matrix: https://gitter.im/ethereum/solidity or https://matrix.to/#/#ethereum_solidity:gitter.im.

#### <img src="https://avatars.githubusercontent.com/u/61954243?u=092f34588505a68339fa458522081fc3c6a05dcb&v=4" width="50">[MartinSchere](https://github.com/MartinSchere) commented at [2021-07-26 15:30](https://github.com/ethereum/solidity/issues/11692#issuecomment-886805548):

Yes, exactly. `balanceBefore` and `balanceAfter` are different, which is why I get a revert when trying to call `transferFrom` after the last line of the function.

I found a workaround by moving the first `transferFrom` below the `checkLP` function described in the example. Now the balance is correct. I would still consider this a potentially dangerous bug.

I asked for help in gitter, but I got your answer here first.

#### <img src="https://avatars.githubusercontent.com/u/13174375?u=52d702cb6bec53b561afa293cf9cd53ef7a63924&v=4" width="50">[hrkrshnn](https://github.com/hrkrshnn) commented at [2021-07-26 15:40](https://github.com/ethereum/solidity/issues/11692#issuecomment-886812557):

@MartinSchere I don't really follow the problem. Could you please contact me on `hrkrshnn:matrix.org` on gitter / matrix? Or just ping me on solidity gitter?

#### <img src="https://avatars.githubusercontent.com/u/61954243?u=092f34588505a68339fa458522081fc3c6a05dcb&v=4" width="50">[MartinSchere](https://github.com/MartinSchere) commented at [2021-07-26 16:28](https://github.com/ethereum/solidity/issues/11692#issuecomment-886850335):

Sure

#### <img src="https://avatars.githubusercontent.com/in/15368?v=4" width="50">[github-actions](https://github.com/apps/github-actions) commented at [2023-02-07 12:04](https://github.com/ethereum/solidity/issues/11692#issuecomment-1420664275):

Hi everyone! This issue has been closed due to inactivity.
If you think this issue is still relevant in the latest Solidity version and you have something to [contribute](https://docs.soliditylang.org/en/latest/contributing.html), feel free to reopen.
However, unless the issue is a concrete proposal that can be implemented, we recommend starting a language discussion on the [forum](https://forum.soliditylang.org) instead.


-------------------------------------------------------------------------------



[Export of Github issue for [ethereum/solidity](https://github.com/ethereum/solidity). Generated on 2024.12.15 at 06:45:24.]
